### PR TITLE
ethapi: add state override to estimateGas

### LINKED ISSUE #27800: Allow state override on eth_estimateGas
# Rationale

Currently, you can not perform an `eth_estimateGas` on something that isn't in the 'correct' state as the evm will, correctly, return an error. For example, on an erc20 you may wish to send an `approve` followed by a `transfer` call. You can estimate gas for the `approve` but not the `transfer`. 

# Implementation

Add an additional parameter to the end of the request in the same way `eth_call` does. 
Are you willing to implement this feature? Yes

### PR DESCRIPTION
**Pull request by amintalebi**
Fixes #27800

### CODE REVIEW

**Comment by s1na** [comment]
Please update the number of parameters in `internal/web3ext/web3ext.go`. Otherwise the overrides won't be recognized in the console.

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -1176,6 +1176,10 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
 		if err != nil {
 			return 0, err
 		}
+		err = overrides.Apply(state)
```
Comment (jsvisa) [inline internal/ethapi/api.go:1179]: overrides maybe nil, we should check before use.

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -1134,7 +1134,7 @@ func (s *BlockChainAPI) Call(ctx context.Context, args TransactionArgs, blockNrO
 	return result.Return(), result.Err
 }
 
-func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash rpc.BlockNumberOrHash, gasCap uint64) (hexutil.Uint64, error) {
+func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash rpc.BlockNumberOrHash, overrides *StateOverride, gasCap uint64) (hexutil.Uint64, error) {
```
Comment (jsvisa) [inline internal/ethapi/api.go:1137]: prefer to add this arg in the ending

Comment (s1na) [reply internal/ethapi/api.go:1137]: Yes this is a must. Otherwise it becomes a backwards-incompatible change.

File: internal/ethapi/api.go (hunk 1 above)
Comment (amintalebi) [reply internal/ethapi/api.go:1179]: it gets checked in the `api.go:919`

File: internal/ethapi/api.go (hunk 2 above)
Comment (amintalebi) [reply internal/ethapi/api.go:1137]: This function is not the actual external api. In the external api it's the last argument. I added it in the same order with `doCall` for this function.

**Review by s1na — APPROVED** [review]
LGTM thanks

**Pull request merged**